### PR TITLE
Source based coverage

### PR DESCRIPTION
**Pull request by Shadoom7**
Add a source-based coverage build.

### CODE REVIEW

File: experiment/run_coverage.py [hunk 1]
```diff
@@ -69,4 +69,4 @@ def do_coverage_run(  # pylint: disable=too-many-locals
                          'coverage_binary': coverage_binary,
                          'output': result.output[-new_process.LOG_LIMIT_FIELD:],
                      })
-    return find_crashing_units(crashes_dir)
```
[the remark this answers is not in the document]
Comment (Shadoom7) [reply experiment/run_coverage.py]: It's mistyped, have already corrected it.

File: common/test_new_process.py [hunk 2]
```diff
@@ -45,7 +45,7 @@ def test_timeout(self):
         start_time = time.time()
         result = new_process.execute(self.COMMAND, timeout=.1)
         end_time = time.time()
-        assert end_time - start_time < 1
+        assert end_time - start_time < 5
```
[the remark this answers is not in the document]
Comment (Shadoom7) [reply common/test_new_process.py:48]: Sure, I've done it.

File: fuzzers/coverage_source_based/fuzzer.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+import os
+
+from fuzzers import utils
+
+
+def build():
+    """Build benchmark."""
+    cflags = [
+        '-fprofile-instr-generate', '-fcoverage-mapping', '-stdlib=libstdc++'
```
Comment (jonathanmetzman) [inline fuzzers/coverage_source_based/fuzzer.py:24]: I'm surprised you need to link against libstdc++ for this?
Could you post one of the errors you see when you didn't do this?
(you can use gists.github.com if it is very long)

Comment (Shadoom7) [reply fuzzers/coverage_source_based/fuzzer.py:24]: It looks like this:
/usr/bin/ld: /usr/lib/libFuzzer.a(FuzzerFork.o): undefined reference to symbol '_ZNSt18condition_variable10notify_oneEv@@GLIBCXX_3.4.11'
//usr/lib/x86_64-linux-gnu/libstdc++.so.6: error adding symbols: DSO missing from command line

Comment (jonathanmetzman) [reply fuzzers/coverage_source_based/fuzzer.py:24]: OK, could you look at how libFuzzer is compiled in fuzzers/libfuzzer/builder.Dockerfile use that method and then try getting rid of libstdc++ here?

Comment (Shadoom7) [reply fuzzers/coverage_source_based/fuzzer.py:24]: Sure!

**Comment by jonathanmetzman** [comment]
CI doesn't test coverage builds. Not great, I need to fix this.

**Comment by jonathanmetzman** [comment]
Let's wait to merge this a second. I'm going to push something directly to this PR so that CI will test coverage builds.

File: experiment/build/test_builder.py [hunk 4]
```diff
@@ -24,7 +24,7 @@
 
 SRC_ROOT = os.path.join(os.path.dirname(__file__), os.pardir, os.pardir)
 
-FUZZER_BLACKLIST = {'coverage'}
+FUZZER_BLACKLIST = {'coverage', 'coverage_source_based'}
```
Comment (lszekeres) [inline experiment/build/test_builder.py:27]: Consider using `COVERAGE_TOOLS` or `FUZZER_SKIP_LIST`

Comment (Shadoom7) [reply experiment/build/test_builder.py]: Done it. Also updated its name in other files.

**Comment by jonathanmetzman** [comment]
For the record, I think my current "fix" is not working. I think the coverage builds are still not building, they are quitting early like the rest of the unaffected fuzzers (Ruikang: FuzzBench's CI is hacky in that a CI job is started for all "fuzzers" but jobs quit early if the fuzzers aren't changed.

**Comment by Shadoom7** [comment]
> For the record, I think my current "fix" is not working. I think the coverage builds are still not building, they are quitting early like the rest of the unaffected fuzzers (Ruikang: FuzzBench's CI is hacky in that a CI job is started for all "fuzzers" but jobs quit early if the fuzzers aren't changed. Need to debug this :-(

I see, I will look into it.

**Comment by Shadoom7** [comment]
Have undone the CI changes.@jonathanmetzman

**Comment by jonathanmetzman** [comment]
Related to #453

**Pull request merged**